tests/lib/util/source-code.js

```diff
@@ -928,7 +961,110 @@ describe("SourceCode", () => {
             eslint.verify(code, config, "", true);
         });
 
-        it("should not attach duplicate leading comments from previous node", () => {
+        it("should return comments only for children of parent node", () => {
+            const code = [
+                "var foo = {",
+                "    bar: 'bar'",
+                "    // comment",
+                "};",
+                "var baz;"
+            ].join("\n");
+
+            eslint.reset();
+            eslint.on("Program", assertCommentCount(0, 0));
+            eslint.on("VariableDeclaration", assertCommentCount(0, 0));
+            eslint.on("VariableDeclerator", assertCommentCount(0, 0));
+            eslint.on("Identifier", assertCommentCount(0, 0));
+            eslint.on("ObjectExpression", assertCommentCount(0, 0));
+            eslint.on("Property", assertCommentCount(0, 1));
+            eslint.on("Literal", assertCommentCount(0, 0));
+
+            eslint.verify(code, config, "", true);
+        });
+
+        it("should return comments for an export default anonymous class", () => {
+            const code = [
+                "/**",
+                " * this is anonymous class.",
+                " */",
+                "export default class {",
+                "    /**",
+                "     * this is method1.",
+                "     */",
+                "    method1(){",
+                "    }",
+                "}"
+            ].join("\n");
+
+            eslint.reset();
+            eslint.on("Program", assertCommentCount(0, 0));
+            eslint.on("ExportDefaultDeclaration", assertCommentCount(1, 0));
+            eslint.on("ClassDeclaration", assertCommentCount(0, 0));
+            eslint.on("ClassBody", assertCommentCount(0, 0));
+            eslint.on("ClassBody", assertCommentCount(0, 0));
+            eslint.on("MethodDefinition", assertCommentCount(1, 0));
+            eslint.on("Identifier", assertCommentCount(0, 0));
+            eslint.on("FunctionExpression", assertCommentCount(0, 0));
+            eslint.on("BlockStatement", assertCommentCount(0, 0));
+
+            eslint.verify(code, config, "", true);
+        });
+
+        it("should return leading comments", () => {
+            let varDeclCount = 0;
+            const code = [
+                "//#!/usr/bin/env node",
+                "var a;",
+                "// foo",
```

Comment:
Does this comment count as trailing for `var a;` *and* leading for `var b;`?

I don't think this is a problem since people who want to iterate over all comments can just iterate over the comment store without worrying about attachment, but I also want to make sure I understand what's going on here.

Maybe some comments around the asserts would help? (E.g., `assertCommentCount(1, 1)(node); // commented shebang, foo`)